### PR TITLE
Draft/Ignore: Reporting aggregated telemetry on solution close

### PR DESCRIPTION
**Pull request by dpoeschl**
Fixes internal TFS bug #1173249

I will **not** be submitting this in its current state. //cc: @heejaechang

### CODE REVIEW

File: src/Workspaces/Core/Portable/Log/FunctionId.cs [hunk 1]
```diff
@@ -78,7 +78,7 @@ internal enum FunctionId
         Workspace_ApplyChanges,
         Workspace_TryGetDocument,
         Workspace_TryGetDocumentFromInProgressSolution,
-        Workspace_Solution_LinkedFileDiffMergingSession,
+        Workspace_Solution_LinkedFileDiffMergingSession, // Not in use
```
Comment (dpoeschl) [inline src/Workspaces/Core/Portable/Log/FunctionId.cs]: @heejaechang I have to leave this one in the list to preserve numbering, right?

File: src/Workspaces/Core/Portable/Telemetry/TelemetryAggregationCoordinator.cs [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+            {
+                workspaceServices.Workspace.WorkspaceChanged += OnWorkspaceChanged;
+            }
+
+            public void RegisterSolutionClosedLogger(IAggregatedTelemetryLogger logger)
+            {
+                loggers.Add(logger);
+            }
+
+            private void OnWorkspaceChanged(object sender, WorkspaceChangeEventArgs args)
```
Comment (dpoeschl) [inline src/Workspaces/Core/Portable/Telemetry/TelemetryAggregationCoordinator.cs]: @heejaechang  I don't get notified of anything here when the user just closes VS with a solution still open. I think we may need to switch implementations to one where the Workspace has special knowledge of this service and can notify it on Dispose (or something)

**Comment by dpoeschl** [comment]
@heejaechang Any initial thoughts would be appreciated. I've also left a couple comments for you inline that relate to our discussion about this the other day.

So far, this implementation appears to work for cases where the solution is closed explicitly (not implicitly during VS shutdown).

File: src/Workspaces/Core/Portable/LinkedFileDiffMerging/LinkedFileDiffMergingAggregatedLogger.cs [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+        private const string TotalLinkedFileGroupsProcessed = "TotalLinkedFileGroupsProcessed";
+        private const string TotalIdenticalDiffs = "TotalIdenticalDiffs";
+        private const string TotalIsolatedDiffs = "TotalIsolatedDiffs";
+        private const string TotalOverlappingDistinctDiffs = "TotalOverlappingDistinctDiffs";
+        private const string TotalOverlappingDistinctDiffsWithSameSpan = "TotalOverlappingDistinctDiffsWithSameSpan";
+        private const string TotalOverlappingDistinctDiffsWithSameSpanAndSubstringRelation = "TotalOverlappingDistinctDiffsWithSameSpanAndSubstringRelation";
+        private const string TotalInsertedMergeConflictComments = "TotalInsertedMergeConflictComments";
+        private const string TotalInsertedMergeConflictCommentsAtAdjustedLocation = "TotalInsertedMergeConflictCommentsAtAdjustedLocation";
+
+        public static KeyValueLogMessage Create(int sessionId, LinkedFileDiffMergingAggregatedLogger aggregateInfo)
```
Comment (dpoeschl) [inline src/Workspaces/Core/Portable/LinkedFileDiffMerging/LinkedFileDiffMergingAggregatedLogger.cs]: This is a new structure with new keys, etc. Do I need to define this externally, or does the KeyValueLogMessage make everything just work on the server side?

Comment (heejaechang) [inline src/Workspaces/Core/Portable/LinkedFileDiffMerging/LinkedFileDiffMergingAggregatedLogger.cs]: this should be fine as it is.

File: src/Workspaces/Core/Portable/LinkedFileDiffMerging/LinkedFileDiffMergingAggregatedLogger.cs [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+            }
+        }
+
+        internal void LogSession(LinkedFileDiffMergingSessionInfo sessionInfo)
+        {
+            lock (instanceLock)
+            {
+                totalSessions++;
+
+                if (sessionInfo.LinkedFileGroups.Any())
```
Comment (heejaechang) [inline src/Workspaces/Core/Portable/LinkedFileDiffMerging/LinkedFileDiffMergingAggregatedLogger.cs]: Count > 1 ?

File: src/Workspaces/Core/Portable/LinkedFileDiffMerging/LinkedFileDiffMergingAggregatedLogger.cs [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+                        totalOverlappingDistinctDiffs += groupInfo.OverlappingDistinctDiffs;
+                        totalOverlappingDistinctDiffsWithSameSpan += groupInfo.OverlappingDistinctDiffsWithSameSpan;
+                        totalOverlappingDistinctDiffsWithSameSpanAndSubstringRelation += groupInfo.OverlappingDistinctDiffsWithSameSpanAndSubstringRelation;
+                        totalInsertedMergeConflictComments += groupInfo.InsertedMergeConflictComments;
+                        totalInsertedMergeConflictCommentsAtAdjustedLocation += groupInfo.InsertedMergeConflictCommentsAtAdjustedLocation;
+
+                        if (groupInfo.InsertedMergeConflictComments > 0 ||
+                            groupInfo.InsertedMergeConflictCommentsAtAdjustedLocation > 0)
+                        {
+                            groupsWithMergeComments.Add(groupInfo);
```
Comment (heejaechang) [inline src/Workspaces/Core/Portable/LinkedFileDiffMerging/LinkedFileDiffMergingAggregatedLogger.cs]: so, we hold groupsWithMergeComments for entire life time of Solution? why not log it right away?

File: src/Workspaces/Core/Portable/LinkedFileDiffMerging/LinkedFileDiffMergingAggregatedLogger.cs [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+using System.Linq;
+using Microsoft.CodeAnalysis.Internal.Log;
+using Microsoft.CodeAnalysis.Telemetry;
+
+namespace Microsoft.CodeAnalysis
+{
+    internal class LinkedFileDiffMergingAggregatedLogger : IAggregatedTelemetryLogger
+    {
+        private static object globalLock = new object();
+        private static Dictionary<Workspace, LinkedFileDiffMergingAggregatedLogger> loggers = new Dictionary<Workspace, LinkedFileDiffMergingAggregatedLogger>();
```
Comment (heejaechang) [inline src/Workspaces/Core/Portable/LinkedFileDiffMerging/LinkedFileDiffMergingAggregatedLogger.cs]: when does it get cleared?

File: src/Workspaces/Core/Portable/Log/FunctionId.cs (hunk 1 above)
Comment (heejaechang) [inline src/Workspaces/Core/Portable/Log/FunctionId.cs]: you can delete it.

File: src/Workspaces/Core/Portable/Telemetry/TelemetryAggregationCoordinator.cs (hunk 2 above)
Comment (heejaechang) [inline src/Workspaces/Core/Portable/Telemetry/TelemetryAggregationCoordinator.cs]: probably.

File: src/Workspaces/Core/Portable/Telemetry/TelemetryAggregationCoordinator.cs [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+        }
+
+        public IWorkspaceService CreateService(HostWorkspaceServices workspaceServices)
+        {
+            return new TelemetryAggregationCoordinator(workspaceServices);
+        }
+
+        internal class TelemetryAggregationCoordinator : ITelemetryAggregationCoordinator
+        {
+            private readonly IList<IAggregatedTelemetryLogger> loggers = new List<IAggregatedTelemetryLogger>();
```
Comment (heejaechang) [inline src/Workspaces/Core/Portable/Telemetry/TelemetryAggregationCoordinator.cs]: thread-safety?

File: src/Workspaces/Core/Portable/LinkedFileDiffMerging/LinkedFileDiffMergingAggregatedLogger.cs [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+
+        internal static void LogSession(Workspace workspace, LinkedFileDiffMergingSessionInfo sessionInfo)
+        {
+            lock (globalLock)
+            {
+                if (!loggers.ContainsKey(workspace))
+                {
+                    loggers[workspace] = new LinkedFileDiffMergingAggregatedLogger();
+                    var coordinator = workspace.Services.GetService<ITelemetryAggregationCoordinator>();
+                    coordinator.RegisterSolutionClosedLogger(loggers[workspace]);
```
Comment (heejaechang) [inline src/Workspaces/Core/Portable/LinkedFileDiffMerging/LinkedFileDiffMergingAggregatedLogger.cs]: what happen to things like preview workspace, misc workspace and etc?

File: src/Workspaces/Core/Portable/Telemetry/TelemetryAggregationCoordinator.cs [hunk 9]
```diff
@@ -0,0 +1,48 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Composition;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.Host.Mef;
+
+namespace Microsoft.CodeAnalysis.Telemetry
+{
+    [ExportWorkspaceServiceFactory(typeof(ITelemetryAggregationCoordinator), ServiceLayer.Host), Shared]
```
Comment (heejaechang) [inline src/Workspaces/Core/Portable/Telemetry/TelemetryAggregationCoordinator.cs]: you probably need ServiceLayer.Default that doesn't do anything.

File: src/Workspaces/Core/Portable/LinkedFileDiffMerging/LinkedFileDiffMergingAggregatedLogger.cs (hunk 6 above)
Comment (heejaechang) [inline src/Workspaces/Core/Portable/LinkedFileDiffMerging/LinkedFileDiffMergingAggregatedLogger.cs]: there is no one cleaning this up. so ConditionalWeakTable might be better? especially cases such as preview workspace and others? unless you specifically do checking to handle those cases?

File: src/Workspaces/Core/Portable/Log/FunctionId.cs (hunk 1 above)
Comment (heejaechang) [inline src/Workspaces/Core/Portable/Log/FunctionId.cs]: Sorry for late reply David.

Added some comments

**Comment by heejaechang** [comment]
@dpoeschl I added some comment, thanks!

**Pull request closed without merging**